### DISCUSSION TITLE
Example for generating keys

### DISCUSSION THREAD

**Discussion by mm28ajos** [malmeloo/FindMy.py#89] [Q&A]
Hey there,
greate work, thanks a lot! I was just wondering how to operational generate a key pair, store/export it and then use it to fetch reports. Could you give an example / add one to the example folder? Thanks!

**Comment by malmeloo** [comment]
Hey! I'll add a proper example soon, but for now, this should give you an idea:

```py3
>>> import findmy
>>> 
>>> key = findmy.KeyPair.new()
>>> print(key)
KeyPair(name="None", public_key="VLAlUkbtl5UW0nKFt2gJaifiNMHoVkNvVnaBmA==", type=KeyType.UNKNOWN)
>>> 
>>> print(key.private_key_b64)  # this is a string!
CDGVABJiyAgFgDO0/ySK8od3k+me5JJPfOA9nQ==
>>>
>>> key_str = key.private_key_b64 
>>> new_key = findmy.KeyPair.from_b64(key_str)  # import the key
>>>
>>> key == new_key
True
```

You can then use the `KeyPair` object to fetch reports, see the existing examples for details.

**Reply by mm28ajos** [reply]
Thanks!

**Reply by mm28ajos** [reply]
Could you extend the example on how to handle key generation for devices which could roll over keys i.e. change advertisement keys regularly from a fixed pool of keys? How would I fetch reports for a devices with multiple keys?

**Reply by malmeloo** [reply]
For a set of keys that's fixed in advance? There are multiple ways to do that, but the most elegant way would be to extend `RollingKeyPairSource` and implement its methods to always return the same set of keys. An instance of this class can then be passed to the regular `fetch_reports` method, just like a regular `KeyPair`.

I will add it to the list of feature requests, it's not very difficult to implement so it can probably make it into the next release.

**Reply by mm28ajos** [reply]
Yes, I guess that is what I want to achieve, not 100% sure I expressed it correctly, so here same more details: I found [this alternative OpenHaystack firmware](https://github.com/pix/heystack-nrf5x) which allows to flash multiple advertising keys on the chip and then uses one of those keys in a rotating manner to make tracking by third parties more difficult. So I need to fetch reports for all those flashed keys in one query.

There is another [repo](https://github.com/ct-Open-Source/findmy2traccar/blob/main/findmy2traccar.py) I found which does this already but uses the biemster code. See _request_reports_ function there.
